routers/web/projects/workflows.go

```diff
@@ -0,0 +1,624 @@
+	if err != nil {
+		ctx.ServerError("FindWorkflowsByProjectID", err)
+		return
+	}
+	for _, wf := range workflows {
+		wf.Project = p
+	}
+	ctx.Data["Workflows"] = workflows
+
+	ctx.Data["workflowIDStr"] = workflowIDStr
+	var curWorkflow *project_model.Workflow
+	if workflowIDStr == "" { // get first value workflow or the first workflow
+		for _, wf := range workflows {
+			if wf.ID > 0 {
+				curWorkflow = wf
+				break
+			}
+		}
+	} else {
+		workflowID, _ := strconv.ParseInt(workflowIDStr, 10, 64)
+		if workflowID > 0 {
+			for _, wf := range workflows {
+				if wf.ID == workflowID {
+					curWorkflow = wf
+					break
+				}
+			}
+		}
+	}
+	ctx.Data["CurWorkflow"] = curWorkflow
+	ctx.Data["ProjectLink"] = project_model.ProjectLinkForRepo(ctx.Repo.Repository, projectID)
+
+	if p.Type == project_model.TypeRepository {
+		ctx.HTML(200, tmplRepoWorkflows)
+	} else {
+		ctx.HTML(200, tmplOrgWorkflows)
+	}
+}
+
+type WorkflowsPostForm struct {
+	EventID string         `json:"event_id"`
+	Filters map[string]any `json:"filters"`
+	Actions map[string]any `json:"actions"`
+}
+
+// WorkflowsPost handles creating or updating a workflow
+func WorkflowsPost(ctx *context.Context) {
+	projectID := ctx.PathParamInt64("id")
+	p, err := project_model.GetProjectByID(ctx, projectID)
+	if err != nil {
+		if project_model.IsErrProjectNotExist(err) {
+			ctx.NotFound(nil)
+		} else {
+			ctx.ServerError("GetProjectByID", err)
+		}
+		return
+	}
+	if p.Type == project_model.TypeRepository && p.RepoID != ctx.Repo.Repository.ID {
+		ctx.NotFound(nil)
+		return
+	}
+	if (p.Type == project_model.TypeOrganization || p.Type == project_model.TypeIndividual) && p.OwnerID != ctx.ContextUser.ID {
+		ctx.NotFound(nil)
+		return
+	}
+
+	// Handle both form data and JSON data
+	// Handle JSON data
+	form := &WorkflowsPostForm{}
+	content, err := io.ReadAll(ctx.Req.Body)
+	if err != nil {
+		ctx.ServerError("ReadRequestBody", err)
+		return
+	}
+	defer ctx.Req.Body.Close()
+	log.Trace("get " + string(content))
+	if err := json.Unmarshal(content, &form); err != nil {
+		ctx.ServerError("DecodeWorkflowsPostForm", err)
+		return
+	}
+	if form.EventID == "" {
+		ctx.JSON(http.StatusBadRequest, map[string]any{"error": "InvalidEventID", "message": "EventID is required"})
+		return
+	}
+
+	// Convert form data to filters and actions
+	filters := convertFormToFilters(form.Filters)
+	actions := convertFormToActions(form.Actions)
+
+	// Validate: at least one action must be configured
+	if len(actions) == 0 {
+		ctx.JSON(http.StatusBadRequest, map[string]any{
+			"error":   "NoActions",
+			"message": ctx.Tr("projects.workflows.error.at_least_one_action"),
+		})
+		return
+	}
+
+	eventID, _ := strconv.ParseInt(form.EventID, 10, 64)
+	if eventID == 0 {
+		// check if workflow event is valid
+		if !project_model.IsValidWorkflowEvent(form.EventID) {
+			ctx.JSON(http.StatusBadRequest, map[string]any{"error": "EventID is invalid"})
```

Comment:
Why sometimes `"error": "InvalidEventID",` sometimes `"error": "EventID is invalid"`? Can you use consistency designs?

And there is already JSONError to help to avoid low-level mistakes.